### PR TITLE
add isSelfJoinAllowed configurable flag for queues

### PR DESCRIPTION
**Pull request by daltonfury42**
We want to try keep self-join for queues (by link) disabled by default. In the proposed new flow, the queue manager would scan generate the QR code and put it all around the vaccination centre.

**Flow 1: Default**

1. The queue manager adds the person manually to the queue by asking for his name and contact number.
2. He would tell the person to wait outside, and to check the status via QR or link send my SMS (SMS will be enabled manually for accounts of covid vaccination centres and hospitals enrolled with us).
3. When the user scans the QRs, it will ask for his mobile number and take him to the token status page.

**Flow 2: Self Join enabled.**

If the admin wants to let people join own their own via the queue link / QR, it will be a configurable queue setting. If admin has enabled "Self Join", upon scanning the QR, it will ask for mobile number as usual. If the mobile number is already in the queue, he will be taken to the token status page, else he will be asked his name, and he can join.

Adding a self join flag and check when creating a token.

More on this feature discussed [here](https://github.com/SimplQ/simplQ-frontend/issues/621).

### CODE REVIEW

**Comment by daltonfury42** [comment]
I have to add the db migration script. 🤔

File: simplq/src/main/java/me/simplq/service/TokenService.java [hunk 1]
```diff
@@ -99,6 +99,9 @@ public TokenDetailResponse createToken(CreateTokenRequest createTokenRequest) {
                     throw SQInvalidRequestException.queueDeletedException();
                   } else if (queue.isFull()) {
                     throw SQInvalidRequestException.queueIsFullException();
+                  } else if (queue.isSelfJoinAllowed() || !loggedInUserInfo.getUserId()
```
Comment (chalx) [inline simplq/src/main/java/me/simplq/service/TokenService.java]: This should be extracted into other methods and use function composition. It is very hard to follow.
```java
  Predicate<Queue> isNotPaused() {
    return queue -> {
      if (queue.getStatus() == QueueStatus.PAUSED) {
        throw SQInvalidRequestException.queuePausedException();
      }

      return true;
    };
  }

  Predicate<Queue> isNotDeleted() {
    return queue -> {
      if (queue.getStatus() == QueueStatus.DELETED) {
        throw SQInvalidRequestException.queueDeletedException();
      }

      return true;
    };
  }

  Predicate<Queue> isNotFull() {
    return queue -> {
      if (queue.isFull()) {
        throw SQInvalidRequestException.queueDeletedException();
      }

      return true;
    };
  }

  void validateQueue(Queue queue) {
    isNotPaused()
        .and(isNotDeleted())
        .and(isNotFull())
        .test(queue);
  }
```
and then just call ```validateQueue(queue)```

Comment (chalx) [reply simplq/src/main/java/me/simplq/service/TokenService.java]: Also we can extract validations in an validation class and have the code in one place.

File: simplq/src/main/java/me/simplq/controller/model/queue/PatchQueueRequest.java [hunk 2]
```diff
@@ -4,5 +4,7 @@
 
 @Data
```
[the remark this answers is not in the document]
Comment (daltonfury42) [reply simplq/src/main/java/me/simplq/controller/model/queue/PatchQueueRequest.java:5]: Done

File: simplq/src/main/java/me/simplq/service/TokenService.java (hunk 1 above)
Comment (daltonfury42) [reply simplq/src/main/java/me/simplq/service/TokenService.java]: lovely. Thanks. Will do this tomorrow morning.

Comment (daltonfury42) [reply simplq/src/main/java/me/simplq/service/TokenService.java]: @chalx I will do this as a separate PR.

File: simplq/src/main/java/me/simplq/service/QueueService.java [hunk 3]
```diff
@@ -179,23 +181,30 @@ public QueueStatusResponse getQueueStatusByName(String queueName) {
                         .count(),
                     Long.valueOf(queue.getTokens().size()),
                     queue.getSlotsLeft(),
-                    queue.getQueueCreationTimestamp()))
+                    queue.getQueueCreationTimestamp(),
```
[the remark this answers is not in the document]
Comment (daltonfury42) [reply simplq/src/main/java/me/simplq/service/QueueService.java:184]: Will take this also as a separate PR.
